{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "# Generate statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "## Run this for both HiRES or CHARM experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "-- \u001b[1mAttaching core tidyverse packages\u001b[22m ------------------------ tidyverse 2.0.0 --\n",
      "\u001b[32mv\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32mv\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.4\n",
      "\u001b[32mv\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32mv\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32mv\u001b[39m \u001b[34mggplot2  \u001b[39m 3.4.4     \u001b[32mv\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32mv\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32mv\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.0\n",
      "\u001b[32mv\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "-- \u001b[1mConflicts\u001b[22m ------------------------------------------ tidyverse_conflicts() --\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31mx\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mi\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Warning message:\n",
      "\"\u001b[1m\u001b[22m`read_table2()` was deprecated in readr 2.0.0.\n",
      "\u001b[36mi\u001b[39m Please use `read_table()` instead.\"\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  gene = \u001b[31mcol_character()\u001b[39m,\n",
      "  R1P10013 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  gene = \u001b[31mcol_character()\u001b[39m,\n",
      "  R1P10013 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "cols(\n",
      "  X1 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X2 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X3 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X4 = \u001b[31mcol_character()\u001b[39m,\n",
      "  X5 = \u001b[32mcol_double()\u001b[39m\n",
      ")\n",
      "\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m1\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m2\u001b[39m\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\u001b[32mdbl\u001b[39m (1): X2\n",
      "\n",
      "\u001b[36mi\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mi\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m1\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m2\u001b[39m\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (1): X1\n",
      "\u001b[32mdbl\u001b[39m (1): X2\n",
      "\n",
      "\u001b[36mi\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mi\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    }
   ],
   "source": [
    "library(tidyverse)\n",
    "library(ggpubr)\n",
    "library(yaml)\n",
    "library(patchwork)\n",
    "\n",
    "config <- read_yaml(file = \"../CHARM_preprocess_pipeline/config.yaml\")\n",
    "\n",
    "to_gigabases <- function(raw_bp) {\n",
    "  raw_bp / 4 * 300 / 1e9\n",
    "}\n",
    "\n",
    "extract_sample <- function(path, slice_position, strip_after = NULL) {\n",
    "  sample <- str_split(path, \"/\", simplify = TRUE)[, slice_position]\n",
    "  if (!is.null(strip_after)) {\n",
    "    sample <- str_split(sample, strip_after, simplify = TRUE)[, 1]\n",
    "  }\n",
    "  sample\n",
    "}\n",
    "\n",
    "read_read_stat <- function(path, slice_position = 3, strip_after = NULL) {\n",
    "  read_table2(path, col_names = FALSE) %>%\n",
    "    arrange(X1) %>%\n",
    "    mutate(\n",
    "      sample_id = extract_sample(X1, slice_position, strip_after),\n",
    "      gigabases = to_gigabases(X2)\n",
    "    ) %>%\n",
    "    select(sample_id, gigabases)\n",
    "}\n",
    "\n",
    "read_pair_stat <- function(path, slice_position = 3, suffix_to_remove = NULL) {\n",
    "  read_table2(path, col_names = FALSE) %>%\n",
    "    arrange(X1) %>%\n",
    "    mutate(\n",
    "      sample_id = extract_sample(X1, slice_position),\n",
    "      sample_id = if (!is.null(suffix_to_remove)) str_remove(sample_id, fixed(suffix_to_remove)) else sample_id\n",
    "    ) %>%\n",
    "    select(sample_id, pairs = X2)\n",
    "}\n",
    "\n",
    "raw_reads <- read_read_stat(\"../stat/raw.fq.stat\", strip_after = \"_\") %>%\n",
    "  rename(Rawreads = gigabases)\n",
    "dna_reads <- read_read_stat(\"../stat/dna.fq.stat\") %>%\n",
    "  rename(DNAreads = gigabases)\n",
    "rna_reads <- read_read_stat(\"../stat/rna.fq.stat\") %>%\n",
    "  rename(RNAreads = gigabases)\n",
    "\n",
    "raw_pairs <- read_pair_stat(\"../stat/raw.pairs.stat\") %>%\n",
    "  rename(raw_pairs = pairs)\n",
    "pairs_dedup <- read_pair_stat(\"../stat/pairs.dedup.stat\") %>%\n",
    "  rename(pairs_dedup = pairs)\n",
    "\n",
    "pairs_clean1 <- read_pair_stat(\"../stat/pairs.c1.stat\", slice_position = 5, suffix_to_remove = \".pairs.gz\") %>%\n",
    "  rename(pairs_clean1 = pairs)\n",
    "pairs_clean2 <- read_pair_stat(\"../stat/pairs.c12.stat\", slice_position = 5, suffix_to_remove = \".pairs.gz\") %>%\n",
    "  rename(pairs_clean2 = pairs)\n",
    "pairs_clean3 <- read_pair_stat(\"../stat/pairs.c123.stat\", slice_position = 5, suffix_to_remove = \".pairs.gz\") %>%\n",
    "  rename(pairs_clean3 = pairs)\n",
    "inter_pairs_clean3 <- read_pair_stat(\"../stat/inter.pairs.c123.stat\", slice_position = 5, suffix_to_remove = \".pairs.gz\") %>%\n",
    "  rename(inter_pairs_clean3 = pairs)\n",
    "\n",
    "yperx <- read_table2(\"../stat/yperx.stat\", col_names = FALSE) %>%\n",
    "  arrange(X1) %>%\n",
    "  mutate(sample_id = extract_sample(X1, 2)) %>%\n",
    "  select(sample_id, yperx = X2)\n",
    "\n",
    "stat <- raw_reads %>%\n",
    "  left_join(dna_reads, by = \"sample_id\") %>%\n",
    "  left_join(rna_reads, by = \"sample_id\") %>%\n",
    "  left_join(yperx, by = \"sample_id\") %>%\n",
    "  left_join(raw_pairs, by = \"sample_id\") %>%\n",
    "  left_join(pairs_dedup, by = \"sample_id\") %>%\n",
    "  left_join(pairs_clean1, by = \"sample_id\") %>%\n",
    "  left_join(pairs_clean2, by = \"sample_id\") %>%\n",
    "  left_join(pairs_clean3, by = \"sample_id\") %>%\n",
    "  left_join(inter_pairs_clean3, by = \"sample_id\")\n",
    "\n",
    "rna_gene_counts <- read_table2(\"../result/RNA_Res/counts.gene.total.format.tsv\")\n",
    "rna_gene_matrix <- as.data.frame(rna_gene_counts %>% select(-gene))\n",
    "feature_stat_gene <- tibble(\n",
    "  sample_id = names(rna_gene_matrix),\n",
    "  UMIs_gene = colSums(rna_gene_matrix),\n",
    "  genes_gene = colSums(rna_gene_matrix != 0)\n",
    ")\n",
    "\n",
    "rna_exon_counts <- read_table2(\"../result/RNA_Res/counts.exon.total.format.tsv\")\n",
    "rna_exon_matrix <- as.data.frame(rna_exon_counts %>% select(-gene))\n",
    "feature_stat_exon <- tibble(\n",
    "  sample_id = names(rna_exon_matrix),\n",
    "  UMIs_exon = colSums(rna_exon_matrix),\n",
    "  genes_exon = colSums(rna_exon_matrix != 0)\n",
    ")\n",
    "\n",
    "if (config$if_RNA_snp_split) {\n",
    "  genome1_counts <- read_table2(\"../result/RNA_Res/counts.gene.genome1.tsv\")\n",
    "  genome1_matrix <- as.data.frame(genome1_counts %>% select(-gene))\n",
    "  genome2_counts <- read_table2(\"../result/RNA_Res/counts.gene.genome2.tsv\")\n",
    "  genome2_matrix <- as.data.frame(genome2_counts %>% select(-gene))\n",
    "\n",
    "  feature_stat_genome1 <- tibble(\n",
    "    sample_id = names(genome1_matrix),\n",
    "    UMIs_gene_genome1 = colSums(genome1_matrix),\n",
    "    genes_gene_genome1 = colSums(genome1_matrix != 0)\n",
    "  )\n",
    "\n",
    "  feature_stat_genome2 <- tibble(\n",
    "    sample_id = names(genome2_matrix),\n",
    "    UMIs_gene_genome2 = colSums(genome2_matrix),\n",
    "    genes_gene_genome2 = colSums(genome2_matrix != 0)\n",
    "  )\n",
    "\n",
    "  stat <- stat %>%\n",
    "    left_join(feature_stat_gene, by = \"sample_id\") %>%\n",
    "    left_join(feature_stat_exon, by = \"sample_id\") %>%\n",
    "    left_join(feature_stat_genome1, by = \"sample_id\") %>%\n",
    "    left_join(feature_stat_genome2, by = \"sample_id\")\n",
    "} else {\n",
    "  stat <- stat %>%\n",
    "    left_join(feature_stat_gene, by = \"sample_id\") %>%\n",
    "    left_join(feature_stat_exon, by = \"sample_id\")\n",
    "}\n",
    "\n",
    "stat <- stat %>%\n",
    "  rename(cellname = sample_id)\n",
    "\n",
    "if (config$if_structure) {\n",
    "  rmsd <- read_table2(\"../stat/rmsd.info\", col_names = FALSE) %>%\n",
    "    mutate(\n",
    "      path_parts = str_split(X1, \"/\", simplify = TRUE),\n",
    "      cellname = path_parts[, 1],\n",
    "      res = str_split(path_parts[, 3], fixed(\".\"), simplify = TRUE)[, 2],\n",
    "      TOP3_RMSD = X5\n",
    "    ) %>%\n",
    "    select(cellname, res, TOP3_RMSD) %>%\n",
    "    pivot_wider(names_from = res, values_from = TOP3_RMSD) %>%\n",
    "    arrange(cellname)\n",
    "\n",
    "  stat <- stat %>% left_join(rmsd, by = \"cellname\")\n",
    "}\n",
    "\n",
    "if (config$if_charm) {\n",
    "  charm_reads <- map(\n",
    "    names(config$split),\n",
    "    function(split_name) {\n",
    "      read_csv(paste0(\"../stat/\", split_name, \".read.stat\"), col_names = FALSE) %>%\n",
    "        transmute(\n",
    "          cellname = X1,\n",
    "          !!paste0(split_name, \"_reads\") := X2 / 2 * 300 / 1e9\n",
    "        )\n",
    "    }\n",
    "  ) %>%\n",
    "    reduce(full_join, by = \"cellname\")\n",
    "\n",
    "  stat <- stat %>% full_join(charm_reads, by = \"cellname\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "fill_numeric <- function(x) {\n",
    "  x[is.na(x) | is.nan(x)] <- 0\n",
    "  x\n",
    "}\n",
    "\n",
    "qc_metrics <- stat %>%\n",
    "  mutate(\n",
    "    RNAreadsRatio = RNAreads / (RNAreads + DNAreads),\n",
    "    pairsPerRead = raw_pairs / DNAreads / 1e9 * 300,\n",
    "    pairsValidRatio = pairs_clean3 / raw_pairs,\n",
    "    interPairsRatio = inter_pairs_clean3 / pairs_clean3\n",
    "  ) %>%\n",
    "  mutate(across(where(is.numeric), fill_numeric))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 1 x 25</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>cellname</th><th scope=col>Rawreads</th><th scope=col>DNAreads</th><th scope=col>RNAreads</th><th scope=col>yperx</th><th scope=col>raw_pairs</th><th scope=col>pairs_dedup</th><th scope=col>pairs_clean1</th><th scope=col>pairs_clean2</th><th scope=col>pairs_clean3</th><th scope=col>...</th><th scope=col>1m</th><th scope=col>200k</th><th scope=col>20k</th><th scope=col>50k</th><th scope=col>ct_reads</th><th scope=col>atac_reads</th><th scope=col>RNAreadsRatio</th><th scope=col>pairsPerRead</th><th scope=col>pairsValidRatio</th><th scope=col>interPairsRatio</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>...</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>R1P10013</td><td>3.753487</td><td>3.752322</td><td>0.0010437</td><td>0.005022</td><td>3736619</td><td>779808</td><td>768346</td><td>508560</td><td>508059</td><td>...</td><td>4.232651</td><td>8.206307</td><td>20.8746</td><td>14.853</td><td>0.0063276</td><td>0.007428</td><td>0.0002780704</td><td>0.2987445</td><td>0.1359676</td><td>0.05423189</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 1 x 25\n",
       "\\begin{tabular}{lllllllllllllllllllll}\n",
       " cellname & Rawreads & DNAreads & RNAreads & yperx & raw\\_pairs & pairs\\_dedup & pairs\\_clean1 & pairs\\_clean2 & pairs\\_clean3 & ... & 1m & 200k & 20k & 50k & ct\\_reads & atac\\_reads & RNAreadsRatio & pairsPerRead & pairsValidRatio & interPairsRatio\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & ... & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t R1P10013 & 3.753487 & 3.752322 & 0.0010437 & 0.005022 & 3736619 & 779808 & 768346 & 508560 & 508059 & ... & 4.232651 & 8.206307 & 20.8746 & 14.853 & 0.0063276 & 0.007428 & 0.0002780704 & 0.2987445 & 0.1359676 & 0.05423189\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 1 x 25\n",
       "\n",
       "| cellname &lt;chr&gt; | Rawreads &lt;dbl&gt; | DNAreads &lt;dbl&gt; | RNAreads &lt;dbl&gt; | yperx &lt;dbl&gt; | raw_pairs &lt;dbl&gt; | pairs_dedup &lt;dbl&gt; | pairs_clean1 &lt;dbl&gt; | pairs_clean2 &lt;dbl&gt; | pairs_clean3 &lt;dbl&gt; | ... ... | 1m &lt;dbl&gt; | 200k &lt;dbl&gt; | 20k &lt;dbl&gt; | 50k &lt;dbl&gt; | ct_reads &lt;dbl&gt; | atac_reads &lt;dbl&gt; | RNAreadsRatio &lt;dbl&gt; | pairsPerRead &lt;dbl&gt; | pairsValidRatio &lt;dbl&gt; | interPairsRatio &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| R1P10013 | 3.753487 | 3.752322 | 0.0010437 | 0.005022 | 3736619 | 779808 | 768346 | 508560 | 508059 | ... | 4.232651 | 8.206307 | 20.8746 | 14.853 | 0.0063276 | 0.007428 | 0.0002780704 | 0.2987445 | 0.1359676 | 0.05423189 |\n",
       "\n"
      ],
      "text/plain": [
       "  cellname Rawreads DNAreads RNAreads  yperx    raw_pairs pairs_dedup\n",
       "1 R1P10013 3.753487 3.752322 0.0010437 0.005022 3736619   779808     \n",
       "  pairs_clean1 pairs_clean2 pairs_clean3 ... 1m       200k     20k     50k   \n",
       "1 768346       508560       508059       ... 4.232651 8.206307 20.8746 14.853\n",
       "  ct_reads  atac_reads RNAreadsRatio pairsPerRead pairsValidRatio\n",
       "1 0.0063276 0.007428   0.0002780704  0.2987445    0.1359676      \n",
       "  interPairsRatio\n",
       "1 0.05423189     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "qc_metrics\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "## Run below if this is a CHARM experiment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "  library(Signac)\n",
    "  library(Seurat)\n",
    "  library(EnsDb.Mmusculus.v79)\n",
    "  library(BSgenome.Mmusculus.UCSC.mm10)\n",
    "  library(future)\n",
    "})\n",
    "\n",
    "plan(\"multicore\", workers = 10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "charm <- rna_gene_counts %>%\n",
    "  column_to_rownames(\"gene\") %>%\n",
    "  as.matrix() %>%\n",
    "  CreateSeuratObject(assay = \"rna\", min.cells = 0, min.features = 0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cell_names <- intersect(\n",
    "  rownames(charm@meta.data),\n",
    "  colnames(rna_gene_counts %>% select(-gene))\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "atac_fragments <- CreateFragmentObject(\"../result/fragments/atac.fragments.bgz\", cells = cell_names)\n",
    "ct_fragments <- CreateFragmentObject(\"../result/fragments/ct.fragments.bgz\", cells = cell_names)\n",
    "mm10_genome <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)\n",
    "atac_count_matrix <- GenomeBinMatrix(atac_fragments, binsize = 5000, genome = mm10_genome)\n",
    "ct_count_matrix <- GenomeBinMatrix(ct_fragments, binsize = 5000, genome = mm10_genome)\n",
    "\n",
    "atac_assay <- CreateChromatinAssay(counts = atac_count_matrix, fragments = atac_fragments, genome = \"mm10\")\n",
    "ct_assay <- CreateChromatinAssay(counts = ct_count_matrix, fragments = ct_fragments, genome = \"mm10\")\n",
    "\n",
    "charm[[\"atac\"]] <- atac_assay\n",
    "charm[[\"ct\"]] <- ct_assay\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "charm@meta.data <- charm@meta.data %>%\n",
    "  rownames_to_column(\"cellname\")\n",
    "rownames(charm@meta.data) <- charm@meta.data$cellname\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "qc_metrics <- qc_metrics %>% full_join(\n",
    "  charm@meta.data %>% dplyr::select(cellname, nCount_atac, nCount_ct, TSS.enrichment.atac, TSS.enrichment.ct),\n",
    "  by = \"cellname\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "qc_metrics %>% write_tsv(\"metadata_raw.tsv\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
