{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "\n",
    "plt.rcParams['pdf.use14corefonts'] = True\n",
    "from matplotlib_inline.backend_inline import set_matplotlib_formats\n",
    "set_matplotlib_formats('retina')\n",
    "plt.rcParams['axes.grid'] = False\n",
    "\n",
    "import statsmodels\n",
    "import random\n",
    "import torch\n",
    "\n",
    "def set_all_seeds(seed):\n",
    "    np.random.seed(seed)\n",
    "    random.seed(seed)\n",
    "    torch.manual_seed(seed)\n",
    "    torch.cuda.manual_seed_all(seed)\n",
    "\n",
    "set_all_seeds(42)\n",
    "\n",
    "from sklearn.metrics import roc_curve, auc\n",
    "from joblib import Parallel, delayed\n",
    "import pybedtools\n",
    "from tqdm import tqdm\n",
    "\n",
    "import os\n",
    "os.chdir(\"/shared/zliu/4_cregene/metacell_allcells/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata = pd.read_csv(\"./all.metadata.tsv\", sep=\"\\t\",header=None)\n",
    "metadata.columns = [\"cellname\", \"celltype\",\"mc_id\",\"mc_20\",\"mc_25\"]\n",
    "# majority voting for each mc_id, determine the celltype\n",
    "metadata_mc = metadata.groupby(\"mc_id\")[\"celltype\"].agg(lambda x: x.mode()[0]).reset_index()\n",
    "\n",
    "rnamat = pd.read_csv(\"SCT_metacellmean.tsv.gz\",sep=\"\\t\",index_col=0)\n",
    "rnamat.index.name = \"mc_id\"\n",
    "variable_genes = pd.read_csv(\"./all.top5k.variablegenes.tsv\",header=None).values.T.astype(str)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "genes = pd.read_csv(\"./transcripts.bed\",sep=\"\\t\",header=None)\n",
    "genes.columns = [\"chrom\",\"start\",\"end\",\"name\",\"strand\"]\n",
    "genes[\"start\"] = genes[\"start\"] // 5000 * 5000\n",
    "genes[\"end\"] = genes[\"end\"] // 5000 * 5000 + 5000\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tss = pd.read_csv(\"./tss.bed\",sep=\"\\t\",header=None)\n",
    "tss.columns = [\"chrom\",\"start\",\"end\",\"gene\",\"strand\"]\n",
    "tss[\"genome_coord\"] =  tss.apply(lambda x: f\"{x['chrom']}:{x['start']//5000*5000-int(2e6)}-{x['start']//5000*5000+5000+int(2e6)}\", axis=1)\n",
    "tss = tss.query('chrom != \"chrM\"')[[\"gene\",\"genome_coord\"]].drop_duplicates().reset_index(drop=True)\n",
    "tss[\"tss_id\"] = tss[\"gene\"] + \"_\" + (tss.groupby(\"gene\").cumcount() + 1).astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "considered_genes = pd.read_table(\"./groud_truth/considered_genes.txt\", header=None).values.T.astype(str)[0]\n",
    "tss.query('gene in @considered_genes', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filepaths = glob.glob(\"./corres/*.tsv.gz\")\n",
    "filepaths = [f\"corres/{gene}_1.tsv.gz\" for gene in considered_genes]\n",
    "tss_ids = [x.split(\"/\")[-1].split(\".\")[0] for x in filepaths]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _read_corres(filepath,tss_id):\n",
    "    gene,genome_coord,tss_id = tss.query('tss_id == @tss_id')[[\"gene\",\"genome_coord\",\"tss_id\"]].values[0]\n",
    "    chrom,start_end = genome_coord.split(\":\")\n",
    "    start,end = start_end.split(\"-\")\n",
    "    start,end = int(start),int(end)\n",
    "    tss_bin = start + 2e6\n",
    "\n",
    "    data = pd.read_csv(filepath,sep=\"\\t\")\n",
    "    # if distance < 5, atac_neg10fdr > 2 & obsexp_distance_neg10fdr < -2 than significant\n",
    "    data[\"distance_to_tss\"] = np.arange(-2e6,2e6+5000,5000)\n",
    "    data[\"chrom\"] = chrom\n",
    "    data[\"pos\"] = start + data[\"distance_to_tss\"] + 2e6\n",
    "    data[\"pos\"] = data[\"pos\"].astype(int)\n",
    "    data[\"tss_id\"] = tss_id\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = []\n",
    "for x,tss_id in tqdm(zip(filepaths,tss_ids)):\n",
    "    try:\n",
    "        dfs.append(_read_corres(x,tss_id))\n",
    "    except:\n",
    "        pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.concat(dfs)\n",
    "data[\"gene\"] = data[\"tss_id\"].str.split(\"_\").str[0] \n",
    "print(\"Total tested pairs: \",data.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "data = data.merge(genes[[\"name\",\"start\",\"end\"]],left_on=\"gene\",right_on=\"name\",how=\"left\")\n",
    "data[\"within_gene\"] = data.apply(lambda x: x[\"pos\"] > (x[\"start\"]+5000) and x[\"pos\"] < (x[\"end\"]-5000),axis=1) \n",
    "data_filter_distance = data.query('distance_highexp < 5 or within_gene')\n",
    "data_filter_distance = data_filter_distance.fillna(0)\n",
    "print(\"Total tested pairs: \",data_filter_distance.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import fdrcorrection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter_distance[\"obsexp_distance_fdr\"] = fdrcorrection(data_filter_distance[\"obsexp_distance_pval\"])[1]\n",
    "data_filter_distance[\"atac_fdr\"] = fdrcorrection(data_filter_distance[\"atac_pval\"])[1]\n",
    "data_filter_distance[\"ct_fdr\"] = fdrcorrection(data_filter_distance[\"ct_pval\"])[1]\n",
    "data_filter_distance[\"distance_fdr\"] = fdrcorrection(data_filter_distance[\"distance_pval\"])[1]\n",
    "\n",
    "data_filter_distance[\"atac_neg10fdr\"] = -np.log10(data_filter_distance[\"atac_fdr\"])\n",
    "data_filter_distance[\"obsexp_distance_neg10fdr\"] = -np.log10(data_filter_distance[\"obsexp_distance_fdr\"])\n",
    "data_filter_distance[\"ct_neg10fdr\"] = -np.log10(data_filter_distance[\"ct_fdr\"])\n",
    "data_filter_distance[\"distance_neg10fdr\"] = -np.log10(data_filter_distance[\"distance_fdr\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(4,4))\n",
    "\n",
    "plt.hist(data_filter_distance[\"obsexp_distance_neg10fdr\"].values,label=\"Obs/Exp Distance\",alpha=0.5,bins = np.arange(0,50,0.5))\n",
    "plt.hist(data_filter_distance[\"ct_neg10fdr\"].values,label=\"H3K27ac\",alpha=0.5,bins = np.arange(0,50,0.5))\n",
    "plt.hist(data_filter_distance[\"atac_neg10fdr\"].values,label=\"Accessibility\",alpha=0.5,bins = np.arange(0,50,0.5))\n",
    "\n",
    "plt.yscale(\"log\")\n",
    "plt.legend()\n",
    "\n",
    "plt.xlabel(\"-log10(FDR)\")\n",
    "plt.ylabel(\"Number of pairs\")\n",
    "\n",
    "plt.xlim(0,45)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# score based on correlation \n",
    "data_filter_distance[\"ct_neg10fdr\"] = data_filter_distance[\"ct_neg10fdr\"].clip(lower=0)\n",
    "data_filter_distance[\"atac_neg10fdr\"] = data_filter_distance[\"atac_neg10fdr\"].clip(lower=0)\n",
    "data_filter_distance[\"obsexp_distance_neg10fdr\"] = data_filter_distance[\"obsexp_distance_neg10fdr\"].clip(lower=0)\n",
    "data_filter_distance = data_filter_distance.reset_index(drop=True)\n",
    "\n",
    "def safe_pow(base, exponent):\n",
    "    \"\"\"处理基数为0且指数为负数的情况,避免除以零错误\"\"\"\n",
    "    return np.where((exponent == -1) & (base == 0), 0, np.power(base, exponent))\n",
    "    \n",
    "data_filter_distance[\"score\"] = (safe_pow(data_filter_distance[\"ct_neg10fdr\"]+1,np.sign(data_filter_distance[\"ct_corr\"])) * safe_pow(data_filter_distance[\"atac_neg10fdr\"]+1,np.sign(data_filter_distance[\"atac_corr\"]))) * np.maximum(1,safe_pow(data_filter_distance[\"obsexp_distance_neg10fdr\"]+1,-np.sign(data_filter_distance[\"obsexp_distance_corr\"])))\n",
    "data_filter_distance[\"score\"] = data_filter_distance[\"score\"] / data_filter_distance.groupby(\"tss_id\")[\"score\"].transform(\"sum\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validation on mESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_table(\"./groud_truth/groud_truth_considered.txt\",header=None)\n",
    "data.columns = [\"chrom\",\"start\",\"end\",\"gene\",\"significant\",\"ref\"]\n",
    "data[\"label\"] = (data[\"significant\"] == True).astype(int)\n",
    "\n",
    "# I cant find the originnal Moorthy2016 paper, so it is not included in the analysis \n",
    "data = data.query('ref != \"Moorthy2016\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "data.to_csv(\"./groud_truth/CHARM_used_groundtruth_raw.tsv\",sep=\"\\t\",index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is a bin/distance based abc score, same nature but different implementation\n",
    "def compute_abc_score(row):\n",
    "    try:\n",
    "        gene = row[\"gene\"]\n",
    "        concat_df_mean = pd.read_csv(f\"./data/{gene}.tsv.gz\", sep=\"\\t\")\n",
    "        concat_df_mean = concat_df_mean.query('cellname.str.startswith(\"P\")', engine='python').\\\n",
    "            drop(columns=[\"cellname\"]).\\\n",
    "            groupby([\"chrom\", \"pos\"]).mean()\n",
    "        concat_df_mean[\"ABC\"] = concat_df_mean[\"atac\"] * concat_df_mean[\"ct\"] / (concat_df_mean[\"distance\"] + 0.1)\n",
    "        chrom = row[\"chrom\"]\n",
    "        start = row[\"start\"] // 5000 * 5000\n",
    "        end = (row[\"end\"] -1) // 5000 * 5000 + 5000\n",
    "        pos = np.arange(start, end, 5000)\n",
    "        return np.max(concat_df_mean.loc[(chrom, pos), \"ABC\"])\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error processing gene {row['gene']}: {e}\")\n",
    "        return np.nan\n",
    "\n",
    "def _merge_rows(df):\n",
    "    gene = df[\"gene\"].values[0]\n",
    "    label = df[\"label\"].values[0]\n",
    "    df_bed = pybedtools.BedTool.from_dataframe(df)\n",
    "    df_bed_merged = df_bed.sort().merge().to_dataframe()\n",
    "    df_bed_merged[\"gene\"] = gene\n",
    "    df_bed_merged[\"label\"] = label\n",
    "    return df_bed_merged\n",
    "\n",
    "roc_data = data.copy()\n",
    "roc_data = roc_data.groupby([\"gene\",\"label\"]).apply(_merge_rows).reset_index(drop=True)\n",
    "roc_data[\"ABC_predict\"] = Parallel(n_jobs=50)(delayed(compute_abc_score)(row[1]) for row in tqdm(roc_data.iterrows()))\n",
    "roc_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# /shared/mwang/CHARM-seq/ABC-Enhancer-Gene-Prediction/input_data_charm/code/pr/4_cor_post_process_AUC_cp.ipynb\n",
    "# find rows in roc_data if it is located in other gene's promoter \n",
    "roc_data_bed = pybedtools.BedTool.from_dataframe(roc_data[[\"chrom\", \"start\", \"end\"]])\n",
    "tss_temp = pd.read_csv(\"./tss.bed\",sep=\"\\t\",header=None)\n",
    "tss_temp.columns = [\"chrom\",\"start\",\"end\",\"gene\",\"strand\"]\n",
    "tss_temp[\"start\"] = tss_temp[\"start\"] - 2500\n",
    "tss_temp[\"end\"] = tss_temp[\"end\"] + 2500\n",
    "tss_temp = tss_temp.query('chrom != \"chrM\" & start > 0')\n",
    "\n",
    "tss_temp = pybedtools.BedTool.from_dataframe(tss_temp[[\"chrom\", \"start\", \"end\"]])\n",
    "\n",
    "roc_data_bed_intersect = roc_data_bed.intersect(tss_temp, u=True).to_dataframe()\n",
    "roc_data_bed_intersect[\"locate_in_promoter\"] = True\n",
    "\n",
    "roc_data = roc_data.merge(roc_data_bed_intersect.drop_duplicates(),\n",
    "                on=[\"chrom\", \"start\", \"end\"],how=\"left\")\n",
    "roc_data[\"locate_in_promoter\"] = roc_data[\"locate_in_promoter\"].fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# /share/home/mwang/shared/mwang/CHARM-seq/ABC-Enhancer-Gene-Prediction/input_data_charm/code/pr/4_cor_post_process_AUC_cp.ipynb\n",
    "\n",
    "# load ABC.score based on Fulco's code https://github.com/broadinstitute/ABC-Enhancer-Gene-Prediction\n",
    "\n",
    "abc_peak = pd.read_table(\"/shared/zliu/4_cregene/metacell_allcells/ABC_score/ABC-Enhancer-Gene-Prediction/results/mESC/Predictions/EnhancerPredictionsAllPutative.tsv.gz\")\n",
    "abc_peak = abc_peak.rename(columns={\"chr\":\"chrom\",\"TargetGene\":\"gene\"})\n",
    "abc_peak[\"pos\"] = (abc_peak[\"start\"] + abc_peak[\"end\"]) // 2 //5000 * 5000\n",
    "\n",
    "abc_peak = abc_peak.groupby([\"chrom\",\"pos\",\"gene\"])[[\"ABC.Score\"]].sum().reset_index()\n",
    "\n",
    "abc_inhouse = roc_data[[\"chrom\",\"start\",\"end\",\"label\",\"ABC_predict\",\"gene\"]]\n",
    "abc_inhouse[\"pos\"] = (abc_inhouse[\"start\"] + abc_inhouse[\"end\"]) // 2 // 5000 * 5000\n",
    "\n",
    "compare_df = abc_inhouse.merge(abc_peak,how=\"left\").merge(data_filter_distance[[\"chrom\",\"pos\",\"gene\",\"score\"]],how=\"left\")\n",
    "compare_df.columns = [\"chrom\",\"start\",\"end\",\"label\",\"ABC_inhouse\",\"gene\",\"pos\",\"ABC_score\",\"Corr_score\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "encode_peak = pd.read_table(\n",
    "    \"/share/home/mwang/shared/mwang/CHARM-seq/ENCODE_rE2G/results/mESC/atac_h3k27ac_intact_hic/encode_e2g_predictions.tsv.gz\"\n",
    ")\n",
    "\n",
    "encode_peak = encode_peak.rename(columns={\"chr\":\"chrom\", \"TargetGene\":\"gene\"})\n",
    "encode_peak[\"pos\"] = (encode_peak[\"start\"] + encode_peak[\"end\"]) // 2 // 5000 * 5000\n",
    "\n",
    "encode_peak = encode_peak.groupby([\"chrom\",\"pos\",\"gene\"])[[\"ENCODE-rE2G.Score\"]].sum().reset_index()\n",
    "\n",
    "compare_df = compare_df.merge(\n",
    "    encode_peak,\n",
    "    how=\"left\",\n",
    "    on=[\"chrom\",\"pos\",\"gene\"]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "compare_df[\"ABC_score\"] = compare_df[\"ABC_score\"].fillna(0)\n",
    "compare_df[\"Corr_score\"] = compare_df[\"Corr_score\"].fillna(0)\n",
    "compare_df[\"ENCODE-rE2G.Score\"] = compare_df[\"ENCODE-rE2G.Score\"].fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GLM regression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "sys.path.append('/shareb/zliu/analysis/')\n",
    "sys.path.append('/shareb/zliu/analysis/CHARMtools')\n",
    "import CHARMtools\n",
    "from CHARMtools import Cell3Ddev as Cell3D\n",
    "from CHARMtools import MultiCell3D\n",
    "from sklearn.cluster import KMeans\n",
    "from CHARMtools import regression\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata = pd.read_csv(\"./all.metadata.tsv\",sep=\"\\t\",header = None)\n",
    "metadata.columns = [\"cellname\",\"celltype\",\"mc_15\",\"mc_20\",\"mc_30\"]\n",
    "\n",
    "with open(\"brain.pkl\", \"rb\") as f:\n",
    "    brain = pickle.load(f)\n",
    "cells = list(brain.cells_dict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnamat = pd.read_csv(\"./all.rna.tsv.gz\",sep=\"\\t\").set_index(\"gene\")\n",
    "rnamat = rnamat.div(rnamat.sum(axis=0), axis=1) * 1e4\n",
    "rnamat = np.log2(rnamat + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def _load_regression_data(gene):\n",
    "    bins_considered = data_filter_distance.query('gene == @gene').query('obsexp_distance_fdr < 0.05 or atac_fdr < 0.05 or ct_fdr < 0.05')[[\"chrom\",\"pos\"]].copy().reset_index(drop=True)\n",
    "\n",
    "    concat_df_mean = pd.read_csv(f\"./data/{gene}.tsv.gz\",sep=\"\\t\")\n",
    "\n",
    "    atac_matrix = concat_df_mean[['chrom','pos','cellname','atac']].pivot_table(index=['chrom','pos'],columns='cellname',values='atac').fillna(0)\n",
    "    ct_matrix = concat_df_mean[['chrom','pos','cellname','ct']].pivot_table(index=['chrom','pos'],columns='cellname',values='ct').fillna(0)\n",
    "    obsexp_contact_matrix = concat_df_mean[['chrom','pos','cellname','obsexp_contact']].pivot_table(index=['chrom','pos'],columns='cellname',values='obsexp_contact').fillna(0)\n",
    "\n",
    "    atac_matrix = atac_matrix.loc[bins_considered.set_index([\"chrom\",\"pos\"]).index,metadata[\"cellname\"].values]\n",
    "    ct_matrix = ct_matrix.loc[bins_considered.set_index([\"chrom\",\"pos\"]).index,metadata[\"cellname\"].values]\n",
    "    obsexp_contact_matrix = obsexp_contact_matrix.loc[bins_considered.set_index([\"chrom\",\"pos\"]).index,metadata[\"cellname\"].values]\n",
    "\n",
    "    atac_matrix_norm = atac_matrix #* metadata[\"atac_FRIP\"].values \n",
    "    ct_matrix_norm = ct_matrix #* metadata[\"ct_FRIP\"].values \n",
    "\n",
    "    # minmax normalization on each row\n",
    "    atac_matrix_norm = atac_matrix_norm.apply(lambda x: (x - x.min()) / (x.max() - x.min()), axis=1)\n",
    "    ct_matrix_norm = ct_matrix_norm.apply(lambda x: (x - x.min()) / (x.max() - x.min()), axis=1)\n",
    "    obsexp_contact_matrix_norm = obsexp_contact_matrix.apply(lambda x: (x - x.min()) / (x.max() - x.min()), axis=1)\n",
    "\n",
    "    # na to 0\n",
    "    atac_matrix_norm = atac_matrix_norm.fillna(0)\n",
    "    ct_matrix_norm = ct_matrix_norm.fillna(0)\n",
    "    obsexp_contact_matrix_norm = obsexp_contact_matrix_norm.fillna(0)\n",
    "\n",
    "    atac_features = atac_matrix_norm.T   \n",
    "    ct_features = ct_matrix_norm.T   \n",
    "    contact_features = obsexp_contact_matrix_norm.T  \n",
    "    X = np.hstack([atac_features, ct_features, contact_features])\n",
    "    y = rnamat.loc[gene, metadata[\"cellname\"].values].values\n",
    "    y = (y  - np.min(y)) / (np.max(y) - np.min(y)) * 100\n",
    "\n",
    "    return X, y, bins_considered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import CHARMtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import sys\n",
    "sys.path.append('/shareb/zliu/analysis/')\n",
    "sys.path.append('/shareb/zliu/analysis/CHARMtools')\n",
    "import CHARMtools\n",
    "from CHARMtools import Cell3Ddev as Cell3D\n",
    "from CHARMtools import MultiCell3D\n",
    "from sklearn.cluster import KMeans\n",
    "from CHARMtools import regression\n",
    "import pickle\n",
    "\n",
    "# rerun this cell if picklable problem occurs\n",
    "def compute_shap(row):\n",
    "    gene = row[\"gene\"]\n",
    "    start = row[\"start\"] // 5000 * 5000\n",
    "    end =   (row[\"end\"] -1) // 5000 * 5000 + 5000\n",
    "    pos = np.arange(start,end, 5000)\n",
    "\n",
    "    set_all_seeds(42)\n",
    "    \n",
    "    regressor = regression.GeneRegressor(alpha=0.1, patience=20, norm_type=\"l2\")\n",
    "    X, y, bins_considered = _load_regression_data(gene)\n",
    "    if X.shape[1] == 0:\n",
    "        return 0,0\n",
    "\n",
    "    model, results = regressor.train_test_model(X=X, y=y, epochs=1000, test_size=0.2, lr=1e-3, batch_size=16, patience=20)\n",
    "    \n",
    "    zscore, shapvalue = regression.set_gene_tile_significance_bootstrapped(\n",
    "        X, y, model.linear.weight.data.numpy(), model.linear.bias.data.numpy(), metadata, \"celltype\", metadata[\"celltype\"].unique()\n",
    "    )\n",
    "    \n",
    "    atac_shapvalue= pd.DataFrame(shapvalue).T.iloc[:, :bins_considered.shape[0]]\n",
    "    atac_shapvalue.columns = bins_considered[\"pos\"].values\n",
    "    ct_shapvalue = pd.DataFrame(shapvalue).T.iloc[:, bins_considered.shape[0]:2*bins_considered.shape[0]]\n",
    "    ct_shapvalue.columns = bins_considered[\"pos\"].values\n",
    "    contact_shapvalue = pd.DataFrame(shapvalue).T.iloc[:, 2*bins_considered.shape[0]:]\n",
    "    contact_shapvalue.columns = bins_considered[\"pos\"].values\n",
    "    \n",
    "    total_shap = atac_shapvalue + ct_shapvalue + contact_shapvalue\n",
    "    pos_to_loc = np.intersect1d(pos, total_shap.columns)\n",
    "    if results[\"test_corr\"] < 0.1:\n",
    "        return 0,results[\"test_corr\"]\n",
    "    else:\n",
    "        return np.max(total_shap.loc[\"mESC\",  pos_to_loc]),results[\"test_corr\"]\n",
    "\n",
    "shapvalues = Parallel(n_jobs=50)(delayed(compute_shap)(row[1]) for row in tqdm(roc_data.iterrows()))\n",
    "\n",
    "roc_data[\"CHARM_shap\"] = [x[0] for x in shapvalues]\n",
    "roc_data[\"CHARM_corr\"] = [x[1] for x in shapvalues]\n",
    "roc_data[\"CHARM_shap\"] = roc_data[\"CHARM_shap\"].fillna(0)\n",
    "roc_data[\"CHARM_corr\"] = roc_data[\"CHARM_corr\"].fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# /share/home/mwang/shared/mwang/CHARM-seq/ABC-Enhancer-Gene-Prediction/input_data_charm/code/pr/4_cor_post_process_AUC.ipynb\n",
    "roc_data_temp = roc_data.reset_index(drop=True).copy()\n",
    "\n",
    "def _calculate_roc_data(true_vec,predict_vec):\n",
    "    fpr, tpr, _ = roc_curve(true_vec, predict_vec)\n",
    "    roc_auc = auc(fpr, tpr)\n",
    "    return fpr, tpr, roc_auc\n",
    "\n",
    "fpr_charm_regression, tpr_charm_regression, roc_auc_charm_regression = _calculate_roc_data(roc_data_temp[\"label\"], roc_data_temp[\"CHARM_shap\"])\n",
    "fpr_abc_inhouse, tpr_abc_inhouse, roc_auc_abc_inhouse = _calculate_roc_data(roc_data_temp[\"label\"], roc_data_temp[\"ABC_predict\"])\n",
    "fpr_abc_fulco, tpr_abc_fulco, roc_auc_abc_fulco = _calculate_roc_data(compare_df[\"label\"], compare_df[\"ABC_score\"])\n",
    "fpr_abc_corscore, tpr_abc_corscore, roc_auc_abc_corscore = _calculate_roc_data(compare_df[\"label\"], compare_df[\"Corr_score\"])\n",
    "fpr_encode, tpr_encode, roc_auc_encode = _calculate_roc_data(\n",
    "    compare_df[\"label\"], compare_df[\"ENCODE-rE2G.Score\"]\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(4, 4))\n",
    "plt.plot(fpr_charm_regression, tpr_charm_regression, color=\"#9e0142\",\n",
    "         label=f\"CHARM Regression (AUC = {roc_auc_charm_regression:.2f})\")\n",
    "plt.plot(fpr_encode, tpr_encode, color=\"#fee08b\", \n",
    "         label=f\"ENCODE rE2G (AUC = {roc_auc_encode:.2f})\")\n",
    "plt.plot(fpr_abc_corscore, tpr_abc_corscore, color=\"#549e3f\",\n",
    "         label=f\"Corr Score (AUC = {roc_auc_abc_corscore:.2f})\")\n",
    "plt.plot(fpr_abc_fulco, tpr_abc_fulco, color=\"#0571b0\",\n",
    "         label=f\"ABC Fulco et al. (AUC = {roc_auc_abc_fulco:.2f})\")\n",
    "\n",
    "plt.plot([0, 1], [0, 1], color=\"grey\", linestyle=\"--\")\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(\"ROC Curve\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.text(0.1, 0.9, f\"n={roc_data_temp.shape[0]}\", fontsize=12, ha='center')\n",
    "\n",
    "\n",
    "threshold = 0.015\n",
    "y_score_charm =  roc_data_temp[\"CHARM_shap\"].values\n",
    "predicted_positive = y_score_charm >= threshold\n",
    "actual_positive = roc_data_temp[\"label\"].values == True\n",
    "actual_negative = roc_data_temp[\"label\"].values == False\n",
    "tp = np.sum(predicted_positive & actual_positive)\n",
    "fp = np.sum(predicted_positive & actual_negative)\n",
    "tpr_at_threshold = tp / np.sum(actual_positive)\n",
    "fpr_at_threshold = fp / np.sum(actual_negative)\n",
    "\n",
    "plt.scatter(fpr_at_threshold, tpr_at_threshold, color=\"red\", zorder=5, label=\"\", s=60, edgecolor='black')\n",
    "\n",
    "plt.legend(loc=\"lower right\", fontsize=9)\n",
    "\n",
    "roc_data_temp = roc_data.reset_index(drop=True).copy()\n",
    "\n",
    "def _calculate_roc_data(true_vec, predict_vec):\n",
    "    fpr, tpr, _ = roc_curve(true_vec, predict_vec)\n",
    "    roc_auc = auc(fpr, tpr)\n",
    "    return fpr, tpr, roc_auc\n",
    "\n",
    "fpr_charm, tpr_charm, roc_auc_charm = _calculate_roc_data(\n",
    "    roc_data_temp[\"label\"], roc_data_temp[\"CHARM_shap\"]\n",
    ")\n",
    "fpr_encode, tpr_encode, roc_auc_encode = _calculate_roc_data(\n",
    "    compare_df[\"label\"], compare_df[\"ENCODE-rE2G.Score\"]\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(4, 4))\n",
    "plt.plot(fpr_charm, tpr_charm, color=\"#9e0142\",\n",
    "         label=f\"CHARM Regression (AUC = {roc_auc_charm:.2f})\")\n",
    "plt.plot(fpr_encode, tpr_encode, color=\"#0571b0\",\n",
    "         label=f\"ENCODE rE2G (AUC = {roc_auc_encode:.2f})\")\n",
    "\n",
    "plt.plot([0, 1], [0, 1], color=\"grey\", linestyle=\"--\")\n",
    "plt.xlabel(\"False Positive Rate\")\n",
    "plt.ylabel(\"True Positive Rate\")\n",
    "plt.title(\"ROC Curve: CHARM vs ENCODE\")\n",
    "plt.legend(loc=\"lower right\")\n",
    "plt.text(0.1, 0.9, f\"n={roc_data_temp.shape[0]}\", fontsize=12, ha='center')\n",
    "\n",
    "threshold = 0.015\n",
    "y_score_charm = roc_data_temp[\"CHARM_shap\"].values\n",
    "predicted_positive = y_score_charm >= threshold\n",
    "actual_positive = roc_data_temp[\"label\"].values == True\n",
    "actual_negative = roc_data_temp[\"label\"].values == False\n",
    "tp = np.sum(predicted_positive & actual_positive)\n",
    "fp = np.sum(predicted_positive & actual_negative)\n",
    "tpr_at_threshold = tp / np.sum(actual_positive)\n",
    "fpr_at_threshold = fp / np.sum(actual_negative)\n",
    "\n",
    "plt.scatter(fpr_at_threshold, tpr_at_threshold, color=\"red\", zorder=5,\n",
    "            label=\"\", s=60, edgecolor='black')\n",
    "\n",
    "plt.legend(loc=\"lower right\", fontsize=9)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# precion recall curve\n",
    "from sklearn.metrics import precision_recall_curve, average_precision_score\n",
    "\n",
    "def _calculate_precision_recall(y_true, y_score):\n",
    "    precision, recall, _ = precision_recall_curve(y_true, y_score)\n",
    "    auprc = average_precision_score(y_true, y_score)\n",
    "    return precision, recall, auprc\n",
    "\n",
    "precision_charm_regression,recall_charm_regression, auprc_charm_regression = _calculate_precision_recall(roc_data_temp[\"label\"], roc_data_temp[\"CHARM_shap\"])\n",
    "precision_abc_inhouse, recall_abc_inhouse, auprc_abc_inhouse = _calculate_precision_recall(roc_data_temp[\"label\"], roc_data_temp[\"ABC_predict\"])\n",
    "precision_abc_fulco, recall_abc_fulco, auprc_abc_fulco = _calculate_precision_recall(compare_df[\"label\"], compare_df[\"ABC_score\"])\n",
    "precision_abc_corscore, recall_abc_corscore, auprc_abc_corscore = _calculate_precision_recall(compare_df[\"label\"], compare_df[\"Corr_score\"])\n",
    "precision_encode, recall_encode, auprc_encode = _calculate_precision_recall(\n",
    "    compare_df[\"label\"], compare_df[\"ENCODE-rE2G.Score\"]\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(4, 4))\n",
    "plt.plot(recall_charm_regression, precision_charm_regression, color=\"#9e0142\",\n",
    "         label=f\"CHARM Regression (AUPRC = {auprc_charm_regression:.2f})\")\n",
    "#plt.plot(recall_abc_inhouse, precision_abc_inhouse, color=\"#fdae61\",\n",
    "#         label=f\"ABC In-house (AUPRC = {auprc_abc_inhouse:.2f})\")\n",
    "plt.plot(recall_encode, precision_encode, color=\"#fee08b\",\n",
    "         label=f\"ENCODE rE2G (AUPRC = {auprc_encode:.2f})\")\n",
    "plt.plot(recall_abc_corscore, precision_abc_corscore, color=\"#549e3f\",\n",
    "         label=f\"Corr Score (AUPRC = {auprc_abc_corscore:.2f})\")\n",
    "plt.plot(recall_abc_fulco, precision_abc_fulco, color=\"#0571b0\",\n",
    "         label=f\"ABC Fulco et al. (AUPRC = {auprc_abc_fulco:.2f})\")\n",
    "\n",
    "plt.xlabel(\"Recall\")\n",
    "plt.ylabel(\"Precision\")\n",
    "plt.title(\"Precision-Recall Curve\")\n",
    "plt.text(0.1, 0.9, f\"n={roc_data_temp.shape[0]}\", fontsize=12, ha='center')\n",
    "\n",
    "threshold = 0.015\n",
    "\n",
    "predicted_positive = roc_data_temp[\"CHARM_shap\"].values >= threshold\n",
    "actual_positive = roc_data_temp[\"label\"].values == True\n",
    "actual_negative = roc_data_temp[\"label\"].values == False\n",
    "\n",
    "tp = np.sum(predicted_positive & actual_positive)\n",
    "fp = np.sum(predicted_positive & actual_negative)\n",
    "recall_at_threshold = tp / np.sum(actual_positive)\n",
    "epsilon = 1e-7\n",
    "precision_at_threshold = tp / (tp + fp + epsilon)\n",
    "plt.scatter(recall_at_threshold, precision_at_threshold,\n",
    "            color=\"red\", zorder=5,                   \n",
    "            label=\"\",s=60,edgecolor='black') \n",
    "\n",
    "plt.legend(loc=\"lower left\", fontsize=9)\n",
    "plt.show()\n",
    "\n",
    "from sklearn.metrics import precision_recall_curve, average_precision_score\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def _calculate_precision_recall(y_true, y_score):\n",
    "    precision, recall, _ = precision_recall_curve(y_true, y_score)\n",
    "    auprc = average_precision_score(y_true, y_score)\n",
    "    return precision, recall, auprc\n",
    "\n",
    "precision_charm_regression, recall_charm_regression, auprc_charm_regression = _calculate_precision_recall(\n",
    "    roc_data_temp[\"label\"], roc_data_temp[\"CHARM_shap\"]\n",
    ")\n",
    "precision_encode, recall_encode, auprc_encode = _calculate_precision_recall(\n",
    "    compare_df[\"label\"], compare_df[\"ENCODE-rE2G.Score\"]\n",
    ")\n",
    "\n",
    "plt.figure(figsize=(4, 4))\n",
    "plt.plot(recall_charm_regression, precision_charm_regression, color=\"#9e0142\",\n",
    "         label=f\"CHARM Regression (AUPRC = {auprc_charm_regression:.2f})\")\n",
    "plt.plot(recall_encode, precision_encode, color=\"#0571b0\",\n",
    "         label=f\"ENCODE rE2G (AUPRC = {auprc_encode:.2f})\")\n",
    "\n",
    "plt.xlabel(\"Recall\")\n",
    "plt.ylabel(\"Precision\")\n",
    "plt.title(\"Precision-Recall Curve\")\n",
    "plt.text(0.1, 0.9, f\"n={roc_data_temp.shape[0]}\", fontsize=12, ha='center')\n",
    "\n",
    "threshold = 0.015\n",
    "predicted_positive = roc_data_temp[\"CHARM_shap\"].values >= threshold\n",
    "actual_positive = roc_data_temp[\"label\"].values == True\n",
    "actual_negative = roc_data_temp[\"label\"].values == False\n",
    "\n",
    "tp = np.sum(predicted_positive & actual_positive)\n",
    "fp = np.sum(predicted_positive & actual_negative)\n",
    "recall_at_threshold = tp / np.sum(actual_positive)\n",
    "epsilon = 1e-7\n",
    "precision_at_threshold = tp / (tp + fp + epsilon)\n",
    "\n",
    "plt.scatter(recall_at_threshold, precision_at_threshold,\n",
    "            color=\"red\", zorder=5,\n",
    "            label=\"\", s=60, edgecolor='black')\n",
    "\n",
    "plt.legend(loc=\"lower left\", fontsize=9)\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from sklearn.metrics import precision_recall_curve\n",
    "\n",
    "precision, recall, thresholds = precision_recall_curve(roc_data_temp[\"label\"], roc_data_temp[\"CHARM_shap\"])\n",
    "f1_scores = 2 * (precision * recall) / (precision + recall + 1e-10)  \n",
    "optimal_idx = np.argmax(f1_scores)\n",
    "optimal_threshold = thresholds[optimal_idx]\n",
    "\n",
    "print(f\"Optimal threshold based on F1-score: {optimal_threshold}\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Graph",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
